api/src/main/java/io/grpc/InternalConfigSelector.java

```diff
@@ -41,14 +42,14 @@
   public static final class Result {
     private final Status status;
     @Nullable
-    private final Object config;
+    private final ConfigOrError config;
```

Comment:
Why not just have `ConfigOrError` instead of Status and `ConfigOrError`? Although one is before parsing the other is after parsing, for the user of `Result`, they don't seem like to check error twice and distinguish between error from `Status` and error from `ConfigOrError`.  

`Result.forError(error)` can just create a result with `Result.config.getError()` being that error.

Reply:
They are two different things. `Result.status` is the status information for ConfigSelector's `selectConfig` operation. It could include error information like failing to find a route. 

Making `Result.config` a `ConfigOrError` is for API completeness. `Result.config` is from the result of `ServiceConfigParser.parseServiceConfig(...)`, which may give a config object or an error. Since the raw config of `Result.config` is generated by us internally, it should never cause ServiceConfigParser to give an error (otherwise, mostly that's a programmer bug). From an API's perspective, I'd prefer to just wire the output of `ServiceConfigParser.parseServiceConfig(...)` to what a `Result` would return to its consumers. Another option is to _assert_ the result of `ServiceConfigParser.parseServiceConfig(...)` is not an error. So `Result.config` will still be an `Object` that guarantees to be a valid config. I don't have a strong opinion regarding these two choices.

Reply:
I think it's better to "enforce" the producer of `Result` to assert it is not an error because of otherwise a programming bug.